### ISSUE TITLE
Docker bridge port mapping failure on same host

### ISSUE THREAD

**Issue by bluefishforsale** [docker/docker#27201]
# Description

Cannot access containers mapped port form same host via public IP.

I can access the mapped ports from other hosts.

```
* From same host
10.10.1.109:8080 from 10.10.1.109  : failure
172.17.0.1:8080 from 10.10.1.109  : success
172.17.0.2:80 from 10.10.1.109  : success
0.0.0.0:8080 from 10.10.1.109  : success

* From another host
10.10.1.109:8080 from 10.10.1.100  : success

```
# Steps to reproduce the issue:

**All curl commands are made from the perspective of the host running the container**
1. docker run -it -p 8080:80 httpd:2.4
2. curl --connect-timeout 10 -I http://10.10.1.109:8080  <-- this is what I need working
3. curl: (28) Connection timed out after 10001 milliseconds
4. curl --connect-timeout 10 -I http://0.0.0.0:8080
   HTTP/1.0 200 OK
   Content-Type: text/html; charset=utf-8
   Content-Length: 2670
   Server: Werkzeug/0.11.11 Python/2.7.12
   Date: Thu, 06 Oct 2016 18:41:32 GMT
5. curl --connect-timeout 10 -I http://127.0.0.1:8080
   HTTP/1.0 200 OK
   Content-Type: text/html; charset=utf-8
   Content-Length: 2670
   Server: Werkzeug/0.11.11 Python/2.7.12
   Date: Thu, 06 Oct 2016 18:41:32 GMT
6. curl --connect-timeout 10 -I http://172.17.0.1:8080
   HTTP/1.0 200 OK
   Content-Type: text/html; charset=utf-8
   Content-Length: 2670
   Server: Werkzeug/0.11.11 Python/2.7.12
   Date: Thu, 06 Oct 2016 18:41:32 GMT
7. docker network inspect bridge

```
[
    {
        "Name": "bridge",
        "Id": "3579e44eba8f28dc7508db5063bbcbd0042e813b74a2d93775cf268cef20f6cd",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.17.0.0/16",
                    "Gateway": "172.17.0.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "d332e25ec14b2b677be29b1c99a17165ea993688fa3f3f5675355cd6a60ca46e": {
                "Name": "peaceful_torvalds",
                "EndpointID": "0e7392c5833523b32538ea699690d3b1d111cea130be2cfbfb5a66d2038711d1",
                "MacAddress": "02:42:ac:11:00:02",
                "IPv4Address": "172.17.0.2/16",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
            "com.docker.network.bridge.name": "docker0",
            "com.docker.network.driver.mtu": "1500"
        },
        "Labels": {}
    }
]

```
1. curl --connect-timeout 10 -I http://172.17.0.2:80
   HTTP/1.1 200 OK
   Date: Thu, 06 Oct 2016 18:49:32 GMT
   Server: Apache/2.4.23 (Unix)
   Last-Modified: Mon, 11 Jun 2007 18:53:14 GMT
   ETag: "2d-432a5e4a73a80"
   Accept-Ranges: bytes
   Content-Length: 45
   Content-Type: text/html
# Describe the results you received:

No connection
# Describe the results you expected:

Successful connection to container port through host bridge network on public IP
# Additional information you deem important (e.g. issue happens only occasionally):

This is for new containers only. Existing containers have working port mapping.

I am able to reproduce this issue on a set of hosts, with different containers, different docker versions, and kernel versions.

New container port mapping DOES work on OS X where I am building / testing containers.
On a set of hosts in a different datacenter with the same docker / kernel versions.
Can't find any significant difference between the working / non working hosts.

failure = 'still can't get to mapped port via public IP"

I've tried adding --userland-proxy=false to the docker exec : result **failure**
I've tried using -p <host_ip>:8080:80 : result **failure**
I've tried adding --privileged to the docker  run command : result **failure**
I've tried manually removing all the IPTABLES rules, and letting docker recreate them: result **failure**
# RELEASE

`CentOS Linux release 7.2.1511 (Core)
`
# UNAME

`Linux webhost5 3.10.0-327.36.1.el7.x86_64 #1 SMP Sun Sep 18 13:04:29 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
`
# DOCKER VERSION

docker version

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

```
# Systemd DOCKER EXEC

```
ExecStart=/usr/bin/docker daemon \
                                 -H tcp://0.0.0.0:2375 \   <-- I tried removing this
                 -H unix:///var/run/docker.sock \
                 --label environment=production \
                                 --userland-proxy=false \
                 --storage-driver=devicemapper \
                 --storage-opt=dm.thinpooldev=/dev/mapper/data-dockerdata \
                 --storage-opt dm.use_deferred_removal=true

```
# IPTABLES NAT

iptables -t nat --list -n

```
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination
DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8          ADDRTYPE match dst-type LOCAL
DOCKER     all  --  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
DOCKER     all  --  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match src-type LOCAL
MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0
MASQUERADE  tcp  --  172.17.0.2           172.17.0.2           tcp dpt:80
MASQUERADE  tcp  --  172.17.0.2           172.17.0.2           tcp dpt:443
MASQUERADE  tcp  --  172.17.0.2           172.17.0.2           tcp dpt:80
MASQUERADE  tcp  --  172.17.0.3           172.17.0.3           tcp dpt:80
MASQUERADE  tcp  --  172.17.0.3           172.17.0.3           tcp dpt:80
MASQUERADE  tcp  --  172.17.0.2           172.17.0.2           tcp dpt:80

Chain DOCKER (3 references)
target     prot opt source               destination
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8080 to:172.17.0.2:80


```
# IPTABLES Filter

iptables  --list -n

```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination
DOCKER-ISOLATION  all  --  0.0.0.0/0            0.0.0.0/0
DOCKER     all  --  0.0.0.0/0            0.0.0.0/0
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain DOCKER (1 references)
target     prot opt source               destination
ACCEPT     tcp  --  0.0.0.0/0            172.17.0.2           tcp dpt:80

Chain DOCKER-ISOLATION (1 references)
target     prot opt source               destination
RETURN     all  --  0.0.0.0/0            0.0.0.0/0


```
# HOST ROUTING TABLE

route -n

```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.10.1.1       0.0.0.0         UG    101    0        0 em1
10.10.1.0       0.0.0.0         255.255.255.0   U     100    0        0 em1
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
```

**Comment by cpuguy83** [comment]
ping @aboch

**Comment by bluefishforsale** [comment]
Any ideas as to what I can try would be really appreciated.

I have exhausted my "maybe it could be this" list of tricks and command modifications.

**Comment by aboch** [comment]
@bluefishforsale 
I was not able to reproduce your issue, starting from a clean nat and filter tables.
My nat table looks like yours, besides that entry for 171.17.0.3 (I am assuming you are running two containers both exposing port 80, but I do not see the corresponding DNAT rule in your o/p). 

Also, what does your filter table look like ?

**Comment by bluefishforsale** [comment]
@aboch A little more in-depth debugging on my part.
Also, i've updated the body of the issue above to include the IP tables filter table, and also my routing table on the host.

I cannot access the local public  ip:port for the container being started FROM the machine itself
I CAN access the public ip:port for the container being started from a DIFFERENT machine

10.10.1.109:8080 from 10.10.1.109  : failure
10.10.1.109:8080 from 10.10.1.100  : success

**Comment by bluefishforsale** [comment]
I've updated description and body to reflect the new information in my last issue update.

**Comment by aboch** [comment]
@bluefishforsale 

I tried to reproduce on a 
`Linux cos0 3.10.0-327.4.5.el7.x86_64 #1 SMP Mon Jan 25 22:07:14 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux` VM.

with `--userland-proxy=false/true`

I still cannot reproduce this issue.

From the attached o/p I do not know what it wrong.

This is a basic use case and we have not heard of this from others. It must have to do with your environment.

You may want to monitor your table counters and identify where the request gets dropped, if it gets dropped by iptables.

I think iptables tracing would help your debugging. I find this post very useful for it: http://backreference.org/2010/06/11/iptables-debugging/

**Comment by aboch** [comment]
@bluefishforsale Did you get a chance to check with iptables tracing if the packet gets dropped by iptables ?

**Comment by aboch** [comment]
I am closing this as there has not been much activity/information exchange and I have not heard of others hitting the same. We can reopen anytime if anybody is hitting this and has time to do more debugging to help identify any root cause.

**Comment by risingnote** [comment]
I am seeing the same problem on Centos 7.3.1611, Docker 17.03.0-ce.

**Comment by cpuguy83** [comment]
@risingnote Need more details, see requested info above.

**Comment by risingnote** [comment]
Sorry for delay, I had to get something deployed so circumvented the problem. 

Going back to the problem now, it looks like this was caused by a misconfigured firewall, however there is a confusing gotcha, which in my case was:

1. IPTables was not configured to allow incoming TCP connections
2. Docker updates IPTables with FORWARD rules (for the bridge network?)
3. Other hosts make TCP connections through the firewall to this host using the FORWARD rules 
4. This host will not connect to itself using the external IP address. I assume the FORWARD rules are not applied.
